pkg/minishift/network/proxy/localproxy.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright (C) 2017 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	goos "os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"syscall"
+
+	"github.com/elazarl/goproxy"
+	"github.com/golang/glog"
+	"github.com/minishift/minishift/pkg/minishift/config"
+	minishiftTLS "github.com/minishift/minishift/pkg/minishift/tls"
+	"github.com/minishift/minishift/pkg/util/os"
+	"github.com/minishift/minishift/pkg/util/os/process"
+)
+
+const (
+	proxyAuthHeader = "Proxy-Authorization"
+)
+
+func StartProxy(proxyPort int, proxyUpstreamAddr string) {
+	// set custom CA
+	setCA(minishiftTLS.CACert, minishiftTLS.CAKey)
+
+	proxy := goproxy.NewProxyHttpServer()
+	proxy.Verbose = true
+	proxy.KeepProxyHeaders = true
+	proxy.OnRequest().HandleConnect(goproxy.AlwaysMitm)
+
+	bindAddr := fmt.Sprintf("0.0.0.0:%d", proxyPort)
+	chain := "" // for logging
+
+	if proxyUpstreamAddr != "" {
+
+		credentials := ""
+		if strings.Index(proxyUpstreamAddr, "@") > -1 {
+			proxyParts := strings.Split(proxyUpstreamAddr, "@")
+			credentials = proxyParts[0]
+
+			if strings.Index(credentials, "http://") > -1 {
+				credentials = strings.TrimPrefix(credentials, "http://")
+				proxyUpstreamAddr = "http://" + proxyParts[1]
+			}
+			if strings.Index(credentials, "https://") > -1 {
+				credentials = strings.TrimPrefix(credentials, "https://")
+				proxyUpstreamAddr = "https://" + proxyParts[1]
+			}
+		}
```

Comment:
@gbraad this complete block you are parsing the URI and then rebuilding it why can't we use https://play.golang.org/p/pc-PjgvnElD one looks much better.